### PR TITLE
add support unix socket for GRPC

### PR DESCRIPTION
**Pull request by junhwong**
### 1. Why is this pull request needed and what does it do?
add support unix socket for GRPC to adapt to some local scenarios that are inconvenient to use TCP

### 4. Does this introduce a backward incompatible change or deprecation?
No

### CODE REVIEW

File: plugin/pkg/parse/host.go [hunk 1]
```diff
@@ -63,6 +63,12 @@ func HostPortOrFile(s ...string) ([]string, error) {
 		}
 
 		if net.ParseIP(stripZone(addr)) == nil {
+			if strings.HasPrefix(host, "unix://") {
+				// TODO 检查是否存在
```
Comment (inigohu) [inline plugin/pkg/parse/host.go]: Could you write it in English?

Comment (junhwong) [reply plugin/pkg/parse/host.go]: ok, changed and commit

File: plugin/pkg/parse/host.go [hunk 2]
```diff
@@ -63,6 +63,12 @@ func HostPortOrFile(s ...string) ([]string, error) {
 		}
 
 		if net.ParseIP(stripZone(addr)) == nil {
+			if strings.HasPrefix(host, "unix://") {
```
Comment (inigohu) [inline plugin/pkg/parse/host.go]: We could include the unix schema in [plugin/pkg/parse/transport.go](https://github.com/coredns/coredns/blob/master/plugin/pkg/parse/transport.go)

File: go.mod [hunk 3]
```diff
@@ -6,7 +6,7 @@ require (
 	github.com/Azure/azure-sdk-for-go v68.0.0+incompatible
 	github.com/Azure/go-autorest/autorest v0.11.28
 	github.com/Azure/go-autorest/autorest/azure/auth v0.5.12
-	github.com/antonmedv/expr v1.12.0
+	github.com/antonmedv/expr v1.12.1
```
Comment (inigohu) [inline go.mod]: why this change?

File: plugin/pkg/parse/host.go [hunk 4]
```diff
@@ -33,6 +33,15 @@ func HostPortOrFile(s ...string) ([]string, error) {
 	var servers []string
 	for _, h := range s {
 		trans, host := Transport(h)
+		if len(host) == 0 {
+			return servers, fmt.Errorf("invalid address: %q", h)
+		}
+
+		switch trans {
```
Comment (inigohu) [inline plugin/pkg/parse/host.go]: Use the `if` statement for a single case.

**Comment by inigohu** [comment]
Could you squash your commits together to make it more concise?

File: go.mod (hunk 3 above)
Comment (junhwong) [reply go.mod]: A misoperation in solving DCO

Comment (inigohu) [reply go.mod]: If this is an unrelated bug fix, please open a new PR

Comment (junhwong) [reply go.mod]: OK, that's no problem. although I don't think it has any impact.

**Comment by junhwong** [comment]
Any other questions? @inigohu

**Comment by inigohu** [comment]
>Any other questions? @inigohu

Please resolve pending comments

File: plugin/pkg/parse/host.go (hunk 4 above)
Comment (junhwong) [reply plugin/pkg/parse/host.go]: see plugin/pkg/parse/host_test.go:66

**Comment by junhwong** [comment]
All resolved.
In addition to `go.mod`, if there is no problem, I will submit a new PR
@inigohu

**Comment by chrisohaver** [comment]
Can you clean the commit history?  You have grabbed an unrelated commit (453f89d70ad9bcf7accf0810f09954fb302fcafe)

**Comment by inigohu** [comment]
Can you include an example of a UNIX socket in `coredns/plugin/grpc/README.md`?

**Comment by junhwong** [comment]
Already added @inigohu 

I don't know how to separate them @chrisohaver

**Comment by junhwong** [comment]
Why use **~~~**  instead of **```** in README.md? 
@inigohu

**Comment by chrisohaver** [comment]
> I don't know how to separate them @chrisohaver

Looking at your fork, I see you are working directly in the master branch ...

Try to do a `rebase` of your master branch to remove the redundant commit.  Try `git rebase -i HEAD~2`, and then drop the unwanted commit.  Then force push master to your fork. This should probably fix things. But git can be confusing.

**Comment by junhwong** [comment]
> 

tks @chrisohaver ,  got it! 


cc @inigohu

**Comment by inigohu** [comment]
>Why use ~~~ instead of ``` in README.md?

They are the same. https://www.markdownguide.org/extended-syntax/#fenced-code-blocks
>Depending on your Markdown processor or editor, you’ll use three backticks (```) or three tildes (~~~)

File: plugin/grpc/README.md [hunk 5]
```diff
@@ -129,7 +129,15 @@ Or with multiple upstreams from the same provider
 }
 ~~~
 
+For local upstreams, can using `UNIX Socket` protocol.
+
+~~~ corefile
+. {
+    grpc . path/to/grpc.sock
```
Comment (chrisohaver) [inline plugin/grpc/README.md]: I think this example would fail to parse correctly. shouldn't it be `grpc . unix:///path/to/grpc.sock`?

File: plugin/grpc/README.md [hunk 6]
```diff
@@ -129,7 +129,15 @@ Or with multiple upstreams from the same provider
 }
 ~~~
 
+For local upstreams, can using `UNIX Socket` protocol.
```
Comment (chrisohaver) [inline plugin/grpc/README.md]:
```suggestion
Forward requests to a local upstream listening on a unix socket `/path/to/grpc.sock`.
```

**Review by chrisohaver — CHANGES_REQUESTED** [review]

File: plugin/grpc/README.md (hunk 5 above)
Comment (junhwong) [reply plugin/grpc/README.md]: No, you're wrong

Grpc is an outer application protocol. The specific communication protocol is unix. They are not a concept.

Comment (junhwong) [reply plugin/grpc/README.md:136]: No, you're wrong

Grpc is an outer application protocol. The specific communication protocol is unix. They are not a concept

Comment (chrisohaver) [reply plugin/grpc/README.md]: I think you're not understanding my comment.  Let me rephrase...

Will your example parse without error?  In your unit test case, the sock path includes the unix uri prefix.  If this example is valid, please add this style of config to the unit tests.

IOW: If the URI prefix is not required, please add a unit test that does not include the unix uri prefix.

Comment (junhwong) [reply plugin/grpc/README.md]: I understand your maind. What I want to express is that what you said is not accurate.
The  `unix://` prefix is necessary, otherwise the resolution will conflict, and the `resolve.conf` definition will be found by default.

Comment (chrisohaver) [reply plugin/grpc/README.md]: > The unix:// prefix is necessary

If it is necessary, your example will fail to work as described.  It would assume the path is to a resolv.conf file, parse it as such, and thus be an invalid example of how to use a unix socket.

Comment (chrisohaver) [inline plugin/grpc/README.md]:
```suggestion
    grpc . unix:///path/to/grpc.sock
```

Comment (chrisohaver) [reply plugin/grpc/README.md]: Please let reviewers resolve conversations. Thanks.

Comment (junhwong) [reply plugin/grpc/README.md]: That was my negligence ...

**Review by junhwong — COMMENTED** [review]
.

**Comment by chrisohaver** [comment]
Curious how you have tested this?

**Comment by junhwong** [comment]
> Curious how you have tested this?

1. Unit tests: eg.  plugin/grpc/setup_test.go
2.  What you mentioned above is only a document issue
3.  This is a simple PR, and it has already worked on my application scenario

**Comment by chrisohaver** [comment]
> it has already worked on my application scenario

This is what I am asking about.  In that scenario, what are you using as the grpc local socket listener?

**Comment by junhwong** [comment]
We use [dns lib](//github.com/miekg/dns) to implementation grpc(`dns.proto`)interface . @chrisohaver

File: plugin/pkg/parse/host.go (hunk 4 above)
Comment (inigohu) [reply plugin/pkg/parse/host.go]: I meant that comparing only the `transport.UNIX` case, it would be better to use an `if` statement instead of `swtich`.

```
if trans == transport.UNIX {
	servers = append(servers, trans+"://"+host)
        continue
}
```

Comment (junhwong) [reply plugin/pkg/parse/host.go]: ...  I misinterpreted it as  `single case test`

Comment (junhwong) [reply plugin/pkg/parse/host.go]: I am accustomed to use `switch`, but i accept your propose and will submit it later.

**Comment by chrisohaver** [comment]
Can you also extend the CoreDNS gRPC server to allow listening on a unix socket?

**Comment by junhwong** [comment]
> Can you also extend the CoreDNS gRPC server to allow listening on a unix socket?

Personally, I think the scene is limited.
CoreDNS is mostly used as a cluster upstream or local cache front-end.

**Comment by chrisohaver** [comment]
> > Can you also extend the CoreDNS gRPC server to allow listening on a unix socket?
> 
> Personally, I think the scene is limited.

As is forwarding gRPC to a local unix socket.

**Comment by chrisohaver** [comment]
I'd like to validate that this feature works as proposed.  Can you recommend to a method for testing it?

**Comment by junhwong** [comment]
> Can you recommend to a method for testing it?

test what?
if it's CoreDNS extend, i don't have a particularly  good way,  you can try two stages:
```
domain-up:unix://path/to { ... }    # for local extend
:1053{
    grpc domain-up unix://path/to 
}
```

**Comment by junhwong** [comment]
extend CoreDNS for support unix socket is also easy，like the example above
`domain-up:unix://path/to { ... }`

I can submit a new PR after this PR is completed, but I still do not recommend extending. There are no instance application scenarios.

**Comment by junhwong** [comment]
@inigohu @chrisohaver 
What is the next process for this PR?

**Comment by chrisohaver** [comment]
> test what?

Your proposed changes. The changes you are requesting we pull into coredns in this pull request. I’d like to test them to make sure they work.

**Comment by junhwong** [comment]
For this PR? They can definitely work!
The unit test can confirm that the configuration has passed, which is enough。
If you want to reconfirm, you can implement a minimum server, the following is my code section, which you can refer to:

```go
func (s *grpcDnsServiceServer) Query(ctx context.Context, in *pb.DnsPacket) (*pb.DnsPacket, error) {
	msg := &dns.Msg{}
	if err := msg.Unpack(in.GetMsg()); err != nil {
		log.Error(err)
		return nil, err
	}
	w := &grpcResponseWriter{
		local: s.local,
	}
	if p, ok := peer.FromContext(ctx); ok {
		w.remote = p.Addr
	}
	s.handler.ServeDNS(w, msg)
	return &pb.DnsPacket{Msg: w.packed}, nil
}
```

```go
listener, err := net.Listen("unix", "path")
server := grpc.NewServer(opts...)
pb.RegisterDnsServiceServer(server, impl)
server.Serve(listener)
```


in `Query` func , print any msg indicates success

**Comment by chrisohaver** [comment]
> For this PR? They can definitely work! 

> The unit test can confirm that the configuration has passed, which is enough

IIUC, The unit tests in your PR merely verify configuration parsing - not functionality.

> ... you can implement a minimum server, the following is my code section, which you can refer to:
> 
> ```go
> func (s *grpcDnsServiceServer) Query(ctx context.Context, in *pb.DnsPacket) (*pb.DnsPacket, error) {
> 	msg := &dns.Msg{}
> 	if err := msg.Unpack(in.GetMsg()); err != nil {
> 		log.Error(err)
> 		return nil, err
> 	}
> 	w := &grpcResponseWriter{
> 		local: s.local,
> 	}
> 	if p, ok := peer.FromContext(ctx); ok {
> 		w.remote = p.Addr
> 	}
> 	s.handler.ServeDNS(w, msg)
> 	return &pb.DnsPacket{Msg: w.packed}, nil
> }
> ```
> 
>  ```go
> listener, err := net.Listen("unix", "path")
> server := grpc.NewServer(opts...)
> pb.RegisterDnsServiceServer(server, impl)
> server.Serve(listener)
> ```


Cool.  Can you integrate this into a test in /test ?

**Comment by junhwong** [comment]
I don't think it's necessary. Testing the configuration is enough

1. Connectivity is done by GRPC itself
2. Protocol correctness is accomplished by the CoreDNS core
3. Combination They were completed by the `Grpc plugin` and have now been completed very well

The test is repeated again, which does not make much sense

**Comment by chrisohaver** [comment]
I'm not comfortable merging this without a functional test.  But if another maintainer is OK with the PR as is, then it can merge.

**Comment by johnbelamaric** [comment]
@chrisohaver agreed. 

Recall, the point of the test being automated is to ensure it still works when things change. It is certainly possible for upstream dependencies to change, and break our functionality. Without a functional automated test, we would not find this when we bump the dependency version, and could release a non-working binary.

**Review by junhwong — COMMENTED** [review]
TBR  @chrisohaver @inigohu

File: plugin/pkg/parse/host.go (hunk 4 above)
Comment (inigohu) [reply plugin/pkg/parse/host.go]: PTAL

Comment (junhwong) [reply plugin/pkg/parse/host.go]: It’s really strange.I already had this modified and submitted it. I even double-checked it.

Comment (junhwong) [reply plugin/pkg/parse/host.go]: @inigohu

File: test/grpc_test.go [hunk 7]
```diff
@@ -56,3 +59,75 @@ func TestGrpc(t *testing.T) {
...
+	msg.Unpack(in.GetMsg())
+	answer := new(dns.Msg)
+	answer.SetRcode(msg, dns.RcodeSuccess)
+	buf, _ := answer.Pack()
+	return &pb.DnsPacket{Msg: buf}, nil
+}
+
+func TestGrpcUnix(t *testing.T) {
+
+	fd := "/tmp/test745260595.grpc"
```
Comment (inigohu) [inline test/grpc_test.go]: Take a look at how [unixsock_test.go](https://go.dev/src/net/unixsock_test.go) does it.
[testUnixAddr](https://go.dev/src/net/mockserver_test.go) uses `os.MkdirTemp` to get a name that is unique

Comment (junhwong) [reply test/grpc_test.go]: ok

File: test/grpc_test.go [hunk 8]
```diff
@@ -56,3 +59,77 @@ func TestGrpc(t *testing.T) {
...
+	buf, _ := answer.Pack()
+	return &pb.DnsPacket{Msg: buf}, nil
+}
+
+func TestGrpcUnix(t *testing.T) {
+	fd, err := os.MkdirTemp("", "*.grpc")
+	if err != nil {
+		t.Fatal(err)
+	}
+	os.Remove(fd)
```
Comment (inigohu) [inline test/grpc_test.go:81]: No need to remove it before use.

File: test/grpc_test.go [hunk 9]
```diff
@@ -56,3 +59,77 @@ func TestGrpc(t *testing.T) {
...
+	msg := &dns.Msg{}
+	msg.Unpack(in.GetMsg())
+	answer := new(dns.Msg)
+	answer.SetRcode(msg, dns.RcodeSuccess)
+	buf, _ := answer.Pack()
+	return &pb.DnsPacket{Msg: buf}, nil
+}
+
+func TestGrpcUnix(t *testing.T) {
+	fd, err := os.MkdirTemp("", "*.grpc")
```
Comment (inigohu) [inline test/grpc_test.go]: `MkdirTemp` creates the directory. It uses an empty pattern to create only the directory and uses `filepath.Join(d, "sock")` to create the file.

File: test/grpc_test.go (hunk 8 above)
Comment (inigohu) [inline test/grpc_test.go]: No need to clean before use

Comment (junhwong) [reply test/grpc_test.go]: PTAL @inigohu

**Comment by inigohu** [comment]
LGTM. 
WDYT @chrisohaver ?

**Comment by chrisohaver** [comment]
> LGTM.
> WDYT @chrisohaver ?

I think it looks good. Thanks for adding the test!

One comment I have is that the organization of the added test is a tad confusing.

The existing test in `grpc_test.go`  (`TestGrpc`) is there to test the grpc _server_ - not the grpc plugin.

The test added (`TestGrpcUnix`) is there to test the grpc _plugin_ .
And the added test arbitrarily uses a grpc listener, which isn't really necessary, and adds a little complexity to the test. That isn't a huge deal, but it makes the goal of the test a bit less clear. I suspect this was just copied from `TestGrpc`.

I think renaming the test to `TestGrpcPluginUnix` would clarify things a little. Maybe also add comment on what is being tested.
And ideally, the coredns server should be listening over normal dns, not grpc.

**Comment by chrisohaver** [comment]
@inigohu, This is a bit off topic, but locally I modified the test in this PR to test concurrent queries ( just for peace of mind and to satisfy some doubts stemming from my ignorance of how grpc and unix domain sockets work).

While I was testing that I noticed that grpc plugin calls `state.Match(ret)`. Under what circumstances would the return not match?  Should we instead check that the Id of the response matches?

**Comment by junhwong** [comment]
Thumbs up for your meticulousness! @chrisohaver 

I have considered where to place it, but this test is not within the scope of a unit test, so it can only be considered as an integration test. Therefore, it is placed here now.

**Comment by junhwong** [comment]
I think there is also ambiguity in TestGrpcPluginUnix. When taken apart, Unix itself becomes a plugin, but this is clearly not the case, so I haven't changed it for now

**Comment by junhwong** [comment]
I just submitted it due to formatting issues during workflow testing. Please take a look again @inigohu @chrisohaver

**Review by chrisohaver — CHANGES_REQUESTED** [review]
Please rename the TestGrpcUbix test to TestGrpcPluginUnix . 
Please add a code comment that describes what is being tested by that test .
In that test, please make the coredns server listen on plain dns, and have the client request over plain dns, not grpc.

**Comment by inigohu** [comment]
>While I was testing that I noticed that grpc plugin calls `state.Match(ret)`. Under what circumstances would the return not match? 

The DNS service could return a malformed response and we should check this.

>Should we instead check that the Id of the response matches?

After reading [this RFC](https://datatracker.ietf.org/doc/html/rfc1035#section-7.3), I would definitely include the ID check in the first place.
>Processing responses
...
The recommended strategy is to do a preliminary matching using the ID
field in the domain header, and then to verify that the question section
corresponds to the information currently desired.

**Comment by junhwong** [comment]
> In that test, please make the coredns server listen on plain dns, and have the client request over plain dns, not grpc.

`plain dns`?  I don't understand what you mean

@chrisohaver

**Comment by chrisohaver** [comment]
> plain dns? I don't understand what you mean

I mean dns over udp, not dns over grpc.

**Comment by junhwong** [comment]
What do you want to test? 
Can you write a simple demo to help me understand?

**Comment by chrisohaver** [comment]
Currently your unit test is structured as follows:

1. A client makes a gRPC dns query to a CoreDNS gRPC server.
2. The _grpc_ plugin proxy forwards the request over gRPC on a unix domain socket ...

Please change the first step to ...

1. A client makes a **UDP** DNS query to a CoreDNS **UDP** server

**Comment by junhwong** [comment]
Why use `UDP DNS`?
I still don't understand what you want to test or achieve.

**Comment by chrisohaver** [comment]
The aim is to simplify the test, removing unnecessary complications that do not improve coverage. Actually an even better way would be to create an instance of the grpc plugin, and call ServeDNS, passing the dns query message to it directly. There are some examples of this for other plugins, eg forward.

**Comment by junhwong** [comment]
Can you give an example or extract some relevant code fragments?

**Comment by chrisohaver** [comment]
TestProxy does this with the _forward_ plugin:

instantiate the plugin (in this example, the plugin is created by parsing a config string, but should also be able to do this by just creating the grpc struct and setting values directly if easier): 

https://github.com/coredns/coredns/blob/604a902e2c7e0317aecaa3666124079c75a31573/plugin/forward/proxy_test.go#L23-L30

create and send a request to the plugin: 

https://github.com/coredns/coredns/blob/604a902e2c7e0317aecaa3666124079c75a31573/plugin/forward/proxy_test.go#L32-L36

**Review by junhwong — COMMENTED** [review]
PTAL @chrisohaver

**Review by chrisohaver — APPROVED** [review]
LGTM. Thanks!

File: plugin/grpc/README.md [hunk 10]
```diff
@@ -129,7 +129,15 @@ Or with multiple upstreams from the same provider
 }
 ~~~
 
+Forward requests to a local upstream listening on a `unix socket`
```
Comment (chrisohaver) [inline plugin/grpc/README.md]:
```suggestion
Forward requests to a local upstream listening on a Unix domain socket.
```

**Comment by junhwong** [comment]
PTAL @chrisohaver @inigohu @jameshartig @dilyevsky @SuperQ @Isolus @greenpau @bradbeam @Tantalor93 @pmoroney @rajansandeep
Thanks!

File: plugin/grpc/README.md [hunk 11]
```diff
@@ -129,7 +129,15 @@ Or with multiple upstreams from the same provider
 }
 ~~~
 
+Forward requests to a local upstream listening on a `unix domain socket`
```
Comment (chrisohaver) [inline plugin/grpc/README.md]:
```suggestion
Forward requests to a local upstream listening on a Unix domain socket.
```

**Review by chrisohaver — CHANGES_REQUESTED** [review]
LGTM, Just minor readme formatting tweaks.  Thanks!

**Review by chrisohaver — APPROVED** [review]

**Review by Tantalor93 — APPROVED** [review]
